app/controllers/product_categories_controller.rb

```diff
@@ -1,22 +1,67 @@
 class ProductCategoriesController < ApplicationController
+  after_filter :last_page
+
+  def last_page
+    session[:previous_page] = request.referer || :back
+  end
+
   def create
-    @product = Product.find(params[:product_id])
-    ProductCategory.create(product_category_params)
-    redirect_to @product
+    product_id = params[:product_id]
+    category_id = params[:product_category][:category_id]
+    @product = Product.find(product_id)
+    unless category_exists_for_product?(product_id, category_id)
+      ProductCategory.create(product_category_params)
+
+      redirect_to @product
+    else
+      flash[:errors] = "You cannot assign the same category"
+
+      redirect_to @product
+    end
   end
 
   def destroy
-    @product = Product.find(params[:product_id])
-    @product_category = ProductCategory.find(params[:id])
-    @product_category.destroy
+    product_category = ProductCategory.find_by(id: params[:id])
+    if product_category
+      category_name = product_category.category.name
+      product_category.destroy
+    end
+    # checks for if destroy goes wrong
```

Comment:
I think the idea here was not a web interface action, but some sort of malicious fiddling. I can't think of how it would work now, so I will remove this check.